linker/pool.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package linker
+
+import "google.golang.org/protobuf/types/descriptorpb"
+
+// allocPool helps allocate descriptor instances. Instead of allocating
+// them one at a time, we allocate a pool -- a large, flat slice to hold
+// all descriptors of a particular kind for a file. We then use capacity
+// in the pool when we need space for individual descriptors.
+type allocPool struct {
+	numMessages   int
+	numFields     int
+	numOneofs     int
+	numEnums      int
+	numEnumValues int
+	numExtensions int
+	numServices   int
+	numMethods    int
```

Comment:
All these `num` fields can be removed in favour of a constructor which allocates the fields. Which also removes the lazy construction below.

Reply:
We can't remove the fields: counting is a recursive operation, so we have to have a struct to track the state/accumulate the counts as we descend through the hierarchy.

Reply:
Could be variables with a closure? Resolving though, as fine as is.